### DISCUSSION TITLE
NoData values / masks

### DISCUSSION THREAD

**Discussion by indus** [libvips/libvips#3437] [Q&A]
I wonder if there is a way to ignore certain values when applying a function like ```rank```.
I have a an image with bad values that are encoded as min or max bit value of the image and I would like to do a 'median' or 'rank' filter that is unaffected by those values. Is that achievable somehow?

**Comment by jcupitt** [comment]
You could do it in two passes perhaps?

Do a median filter to fill in the holes, then do a second median on the result of that. The filled-in holes shouldn't affect the result, since they are already the median value.

**Comment by indus** [comment]
Sorry to bother you again... I'm having trouble to figure out to apply this technique to a gaussianblur. In PS I would select only the good values and blur them (they wouldn't get affected by bad values). Is this possible in vips as well?

**Reply by jcupitt** [reply]
I think I'd fill in the holes with a median filter, then blur the result. If you post some sample images I could write a tiny demo prog.

**Reply by indus** [reply]
While I was trying to come up with some test data I've noticed that PS also has "edge bleeding" on gaussian filter in a selection (There is a description of the problem: https://community.adobe.com/t5/photoshop-ecosystem-discussions/blur-spilling-out-of-bounds/td-p/10150461). My dummy data would look like this: 
![grafik](https://user-images.githubusercontent.com/2046274/232414230-ac300739-bc9a-4aa4-be66-b5b0340d1e16.png)

Here is the artifact I want to get rid of (all arround the edge)
![grafik](https://user-images.githubusercontent.com/2046274/232414489-29cc31fa-14c3-4ba1-911a-864e9992199a.png)

Does this test data work for you?
The actual data is a 16 bit grayscale elevation model with discrete steps. I want to round he the edges only on certain steps. I'm not allowed to make it publicy accessible.

**Reply by jcupitt** [reply]
Hmmm so you want to treat all pixel==255 points as no-data, is that right?

A median filter will only work well for isolated pixels. If you want to mask out a large area, you'll need something fancier. For example, a voronoi fill will the value of the nearest data pixel to each no-data pixel:

https://en.wikipedia.org/wiki/Voronoi_diagram

In libvips, it's called `fill_nearest`:

https://www.libvips.org/API/current/libvips-morphology.html#vips-fill-nearest

So run that, then blur the result, then use your original mask to select the blurred areas.

You'll still get some minor artifacts, but it'll be much better than the photoshop example you have above.
